### ISSUE TITLE
Connection lost on heavy load

### ISSUE THREAD

**Issue by chrisWhyTea** [joowani/python-arango#30]
Hello im playing around with a load testing tool, and trying to crash my application that is using this driver... i test 10k connection attempts with a vareyety of conncurent connections...while handeling 50-100 concurrent connections (based on the arangodb dashboard) without a problem, starting with < 100 concurrent connections is starting to raise  "Error 54,  in Connection reset by peer exceptions" for round about 20% of my connection attempts comming from the HttpClient of the driver... first i thought its a arangodb problem but based on my research arangodb should be able to handle 10k conccurent connections... (if no keep-alive header is send) is there something like a max concurrent connection thing in the driver and if so how to ajust it?

**Comment by bangert** [comment]
what version of arangodb are you using?

I am suddenly seeing something similar on arangodb3-3.1.3

**Comment by chrisWhyTea** [comment]
using 3.1.3

**Comment by joowani** [comment]
Hi @chrisWhyTea,

I believe "Connection reset by peer" errors originate from ArangoDB server, which is dropping the connection for some reason. Please have a look at this thread https://github.com/kennethreitz/requests/issues/3189 and try upgrading your requests[security], openssl, and rebuild your virtualenv (if you have one). Also, could you post the stacktrace if you have any? I have a feeling this is a separate problem from bangert's.

@bangert are you sure that you weren't seeing your issue prior to ArangoDB 3.1.3?

Thanks

**Comment by bangert** [comment]
@joowani not really....  i just upgraded from Fedora 24 to 25, so lots of things changed. 

Thanks for your feedback. i'll open a seperate bug once i have something more concrete... sorry for having highjacked this issue.

**Comment by joowani** [comment]
Ok great thanks @bangert!

**Comment by joowani** [comment]
Merging https://github.com/joowani/python-arango/issues/32 to this issue.

**Comment by joowani** [comment]
I am unable to reproduce the issue. Is this still happening with the latest version of ArangoDB?

**Comment by bangert** [comment]
just for the record: my problem was solved in #29

**Comment by joowani** [comment]
@bangert That's great! I'm closing this issue for now. Feel free to re-open if it's back (preferably reproducible)

**Comment by ysule** [comment]
@joowani I am facing the same issue with ArangoDB 3.2
I am creating the data using PySpark and then sending the dataFrames to the **import_bulk** API of the python-arango Document Class. It works fine when the number of records are < 100K, but my dataset is around 580K in size.
When I use a Stand-alone Python Script, the data size does not matter, but it does not work for my PySpark App. 
Please suggest a workaround.

**Comment by joowani** [comment]
Hi @ysule,

Do you have sample code + data that you could share so I can reproduce this? 

Thanks,
Joohwan

**Comment by ysule** [comment]
@joowani I cannot share the data since it is confidential... I can share the code with you.

3 Vertex Collections, 2 Edge Collections

This code is inserting data to 2 vertex collections but fails for the Third Vertex collection onwards.

[spark-arango-app.txt](https://github.com/joowani/python-arango/files/1351244/spark-arango-app.txt)

**Comment by joowani** [comment]
Thanks for that @ysule. Could you also provide me the full stacktrace? Thanks

**Comment by ysule** [comment]
@joowani it might take sometime... I'll send it soon

**Comment by joowani** [comment]
I did some research and I would like you to try the following (if it's not too much work, please test your code after each step so we can narrow down the problem).

### Step 1
Try updating your security package by running:
```bash
sudo python3 -m pip install "requests[security]"
``` 
or if you use python 2: 
```bash
sudo python -m pip install "requests[security]"
```

### Step 2
Please refer to this [page](http://python-driver-for-arangodb.readthedocs.io/en/master/http.html) and inject your own custom HTTP client. You don't want to use the example code exactly, but instead enable and dial up the keep-alive following this [page](https://github.com/requests/requests/issues/3808). It would look something like this:

```python
# You want something like this in your custom HTTP client
from requests.adapters import HTTPAdapter

class KeepaliveAdapter(HTTPAdapter):
    def init_poolmanager(self, *args, **kwargs):
        kwargs['socket_options'] = HTTPConnection.default_socket_options + [
            (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1),
        ]
        super(KeepaliveAdapter, self).init_poolmanager(*args, **kwargs)

class MyCustomHTTPClient(BaseHTTPClient):

    def __init__(self, retries=5):
        self._session = Session()
        adapter = KeepaliveAdapter()
        self._session.mount('https://', adapter())
        self._session.mount('http://', adapter())
...
```
You may also want to tune the keep-alive of your OS (apparently default in Ubuntu is 6 mins):
```
sysctl -w net.ipv4.tcp_keepalive_time=180
```

### Step 3
Make sure that your connection is not being dropped prematurely. Are you using a proxy or anything like that? How long does the `import_bulk` run for before it throws the exception? Try measuring the operation.

### Step 4 (last resort)
Change your code and run `import_bulk` multiple times on smaller batches.

**Comment by ysule** [comment]
@joowani I changed my code and now instead I read from HDFS directly with my Application. The timeout errors were from Apache Phoenix most probably, since Now the errors that I get are as follows

  **File "/home/daasuser/yayati/spark_arango.py", line 67, in <module>
    load(df,conf)
  File "/home/daasuser/yayati/spark_arango_hdfs.py", line 39, in load
    db.collection(elem['coll_name']).import_bulk(vertexDF.toPandas().to_dict('records'),on_duplicate="update")
  File "/home/daasuser/anaconda2/lib/python2.7/site-packages/arango/api.py", line 22, in wrapped_method
    return conn.handle_request(request, handler)
  File "/home/daasuser/anaconda2/lib/python2.7/site-packages/arango/connection.py", line 164, in handle_request
    return handler(getattr(self, request.method)(**request.kwargs))
  File "/home/daasuser/anaconda2/lib/python2.7/site-packages/arango/collections/standard.py", line 995, in handler
    raise DocumentInsertError(res)
arango.exceptions.DocumentInsertError: [HTTP 413] Payload Too Large**

**Comment by joowani** [comment]
Hi @ysule,

Looks like that error is coming from the ArangoDB server. Have you tried breaking down your input into batches and run import_bulk multiple times?

**Comment by joowani** [comment]
Hi @ysule,

Were you able to resolve your issue?
